File: OS/ref/rPhase1/ref.java
package OS.ref.rPhase1;

import java.io.*;
import java.util.*;

public class ref {

    static List<String[]> M = new ArrayList<>();
    static List<String> Buff = new ArrayList<>();
    static List<String> R = new ArrayList<>();
    static String IR;
    static int IC = 0;
    static boolean C = false;
    static int SI = 0;

    public static void initialize() {
        for (int i = 0; i < 100; i++) {
            M.add(new String[]{"ඞ", "ඞ", "ඞ", "ඞ"});
        }
    }

    public static void mos() {
        if (SI == 1) {
            int i = 0;
            int count = 0;
            while (!Buff.get(0).equals("\n")) {
                i += 1;
            }

            for (int k = 0; k < i; k += 4) {
                M.set(Integer.parseInt(IR.substring(2)) + count, Buff.get(0).substring(k, k + 4).split(""));
                count += 1;
            }
            Buff.remove(0);
        } else if (SI == 2) {
            Buff.add(String.join("", M.get(Integer.parseInt(IR.substring(2)))));
            try (FileWriter out = new FileWriter("output.txt", true)) {
                out.write(Buff.get(0));
                out.write("\n");
            } catch (IOException e) {
                e.printStackTrace();
            }
            Buff.remove(0);
        } else if (SI == 3) {
            System.out.println("\n");
        }
    }

    public static void executeUserPgm() {
        IR = String.join("", M.get(IC));

        if (IR.substring(0, 2).equals("GD")) {
            SI = 1;
            mos();
        } else if (IR.substring(0, 2).equals("PD")) {
            SI = 2;
            mos();
        } else if (IR.substring(0, 2).equals("H")) {
            SI = 3;
            mos();
        } else if (IR.substring(0, 2).equals("LR")) {
            R.add(String.join("", M.get(Integer.parseInt(IR.substring(2)))));
        } else if (IR.substring(0, 2).equals("SR")) {
            M.set(Integer.parseInt(IR.substring(2)), R.toArray(new String[0]));
        } else if (IR.substring(0, 2).equals("CR")) {
            if (R.get(0).equals(String.join("", M.get(Integer.parseInt(IR.substring(2)))))) {
                C = true;
            }
        } else if (IR.substring(0, 2).equals("BT")) {
            if (C) {
                IC = Integer.parseInt(IR.substring(2));
            }
        }
        IC++;
    }

    public static void init() {
        System.out.println("This is init function");
    }

    public static void start_exec() {
        executeUserPgm();
    }

    public static void load() {
        String filename = "input.txt";
        try (BufferedReader file = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = file.readLine()) != null) {
                if (line.startsWith("$AMJ")) {
                    init();
                } else if (line.startsWith("$DTA")) {
                    while (true) {
                        String next_line = file.readLine();
                        if (next_line.startsWith("$END")) {
                            break;
                        }
                        Buff.add(next_line);
                        executeUserPgm();
                        IC++;
                    }
                    String x;
                    do {
                        executeUserPgm();
                        x = String.valueOf(M.get(IC)[0]);
                        IC++;
                    } while (!x.equals("H"));
                } else {
                    int length = line.length();
                    int index = 0;
                    for (int i = 0; i < length; i += 4) {
                        M.set(index, line.substring(i, i + 4).split(""));
                        index++;
                    }
                }
            }
            for (int i = 0; i < 100; i++) {
                System.out.println("M[" + i + "] = " + String.join("", M.get(i)));
            }
        } catch (IOException e) {
            System.out.println("File not found.");
        }
    }

    public static void main(String[] args) {
        initialize();
        load();
    }
}

File: OS/ref/rPhase1/ref.py
'''OS CP'''

M, Buff, R, IR = [],[],[],[]
IC = 0
C = False
SI = 0

def initialize():
    for i in range(100):
        M.append([0,0,0,0])

def mos():
    global Buff
    global IR
    global M
    if SI == 1:
        i = 0
        count = 0
        while Buff[0][i] != '\n':
            i += 1

        for k in range(0,i,4):
            M[int(IR[0][2:])+count] = Buff[0][k:k+4]
            count += 1
        Buff.pop()

    elif SI == 2:
        Buff.append(M[int(IR[0][2:])])
        try:
            with open("/home/aniket/Desktop/SY2/OS/ref/rPhase1/output.txt", 'a') as out:
                out.write(Buff[0])
                out.write("\n")
            out.close()
        except Exception as e:
            print(e)
        Buff.pop()

    elif SI == 3:
        print()
        print()
    IR.pop()

def executeUserPgm():
    global IC
    global SI
    global IR
    global R
    global C
    global M

    IR.append(M[IC])

    if IR[0][:2] == "GD":
        SI = 1
        mos()
    elif IR[0][:2] == "PD":
        SI = 2
        mos()
    elif IR[0][:2] == "H":
        SI = 3
        mos()
    elif IR[0][:2] == "LR":
        R.append(M[int(IR[0][2:])])
        IR.pop()
    elif IR[0][:2] == "SR":
        M[int(IR[2:])] = R
        IR.pop()
    elif IR[0][:2] == "CR":
        if R[0] == M[int(IR[0][2:])]:
            C = True
        IR.pop()
    elif IR[0][:2] == "BT":
        if C is True:
            IC = int(IR[0][2:])
        IR.pop()

def init():
    print("This is init function")

def start_exec():
    executeUserPgm()

def load():
    global Buff
    global IC
    filename = "input.txt"
    try:
        with open(filename, 'r') as file:
            length = 0
            for line in file:
                if line[:4] == "$AMJ":
                    init()
                elif line[:4] == "$DTA":
                    while True:
                        next_line = file.readline()
                        if next_line[:4] == "$END":
                            print()
                            break
                        Buff.append(next_line)
                        executeUserPgm()
                        IC += 1

                    x = ''
                    while x != 'H':
                        executeUserPgm()
                        IC = int(IC) + 1
                        x = M[IC][0]

                else:
                    length = len(line)
                    index = 0

                    for i in range(0,length,4):
                        M[index] = line[i:i+4]
                        index += 1

            for i in range(100):
                print("M["+str(i)+"] = "+str(M[i]))

    except FileNotFoundError:
        print("File not found.")

initialize()